### PR TITLE
Refactor api update

### PR DESCRIPTION
**Pull request by seanlinsley**
https://github.com/modernmsg/modernmsg/issues/472

### CODE REVIEW

File: lib/giftbit/base.rb [hunk 1]
```diff
@@ -0,0 +1,79 @@
+require 'active_support'
+
+begin
+  require 'rest-client'
+rescue LoadError
+end
```
Comment (seanlinsley) [inline lib/giftbit/base.rb]: Seems bad to be silencing this error, since the gem depends on rest-client

File: lib/giftbit/base.rb [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+
+module Giftbit
+  module Base
+    def self.included(klass)
+      klass.extend ClassMethods
+
+      # By default, set the endpoint to production
+      klass.endpoint  ||= 'https://api.giftbit.com/papi/v1/'
+      # By default, set the auth token to be nil
+      klass.auth      ||= ''
```
Comment (seanlinsley) [inline lib/giftbit/base.rb]: This comment is no longer accurate

File: lib/giftbit/base.rb [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+rescue LoadError
+end
+
+module Giftbit
+  module Base
+    def self.included(klass)
+      klass.extend ClassMethods
+
+      # By default, set the endpoint to production
+      klass.endpoint  ||= 'https://api.giftbit.com/papi/v1/'
```
Comment (seanlinsley) [inline lib/giftbit/base.rb]: The previous version set it to the test endpoint by default, which I think makes more sense. That way you don't accidentally create real gift cards when trying out the gem for the first time.

File: lib/giftbit/base.rb [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+          response = resource.send(method, *[options])
+        else
+          response = resource.send(method, *[JSON.generate(body), options])
+        end
+
+        JSON.parse(response)
+      rescue => e
+        if e.response
+          JSON.parse(e.response)
+        end
```
Comment (seanlinsley) [inline lib/giftbit/base.rb]: What about network or SSL exceptions?

File: lib/giftbit/base.rb [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+        resource_options[:headers]['Content-Type'] ||= 'json'
+        response(:post, resource(resource_options)[path], request_options)
+      end
+
+      # PUT (parsed) response from resources
+      def put(path, request_options = {}, resource_options = {})
+        resource_options[:headers] ||= {}
+        resource_options[:headers]['Content-Type'] ||= 'json'
+        response(:put, resource(resource_options)[path], request_options)
+      end
```
Comment (seanlinsley) [inline lib/giftbit/base.rb]: So `put` and `post` require the content type, but `get` and `delete` don't? :-/

File: lib/giftbit/base.rb (hunk 4 above)
Comment (TheBreeze) [inline lib/giftbit/base.rb]: The response method will return nil in such circumstances. We may want to re-raise exceptions that do not have a response, thus delegating the error handling to the application.

**Pull request merged**